server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/event/ComputeMessagePreviewListener.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.james.mailbox.model.MessageResult;
+import org.apache.james.mailbox.store.SessionProvider;
+import org.apache.james.mime4j.dom.Message;
+import org.apache.james.mime4j.stream.MimeConfig;
+import org.apache.james.util.html.HtmlTextExtractor;
+import org.apache.james.util.mime.MessageContentExtractor;
+import org.apache.james.util.mime.MessageContentExtractor.MessageContent;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+
+import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Schedulers;
+
+public class ComputeMessagePreviewListener implements MailboxListener.GroupMailboxListener {
+    public static class ComputeMessagePreviewListenerGroup extends Group {
+
+    }
+
+    static final Group GROUP = new ComputeMessagePreviewListenerGroup();
+
+    private final MessageIdManager messageIdManager;
+    private final MessagePreviewStore messagePreviewStore;
+    private final SessionProvider sessionProvider;
+    private final MessageContentExtractor messageContentExtractor;
+    private final HtmlTextExtractor htmlTextExtractor;
+
+    @Inject
+    public ComputeMessagePreviewListener(SessionProvider sessionProvider, MessageIdManager messageIdManager,
+                                         MessagePreviewStore messagePreviewStore, MessageContentExtractor messageContentExtractor,
+                                         HtmlTextExtractor htmlExtractor) {
+        this.sessionProvider = sessionProvider;
+        this.messageIdManager = messageIdManager;
+        this.messagePreviewStore = messagePreviewStore;
+        this.messageContentExtractor = messageContentExtractor;
+        this.htmlTextExtractor = htmlExtractor;
+    }
+
+    @Override
+    public Group getDefaultGroup() {
+        return GROUP;
+    }
+
+    @Override
+    public void event(Event event) throws MailboxException {
+        if (event instanceof Added) {
+            MailboxSession session = sessionProvider.createSystemSession(event.getUsername());
+            handleAddedEvent((Added) event, session);
+        }
+    }
+
+    private void handleAddedEvent(Added addedEvent, MailboxSession session) throws MailboxException {
+        Flux.fromIterable(messageIdManager.getMessages(getEventMessageIds(addedEvent), FetchGroup.BODY_CONTENT, session))
+            .publishOn(Schedulers.boundedElastic())
+            .map(Throwing.function(messageResult -> Pair.of(messageResult.getMessageId(), computePreview(messageResult))))
+            .flatMap(message -> messagePreviewStore.store(message.getKey(), message.getValue()))
+            .then()
+            .block();
+    }
+
+    private List<MessageId> getEventMessageIds(Added addedEvent) {
+        return addedEvent.getUids()
+            .stream()
+            .map(uid -> addedEvent.getMetaData(uid).getMessageId())
+            .collect(Guavate.toImmutableSet())
+            .asList();
```

Comment:
why do you care to have a List?

Reply:
Because MessageIdManager expectes one.

It should rather accept a collection.

Reply:
Yes `getMessages` from the MessageIdManager expects a list as a param unfortunately